### PR TITLE
Issue#269 Login drop-down menu

### PR DESCRIPTION
**Pull request by igypigi**
Added login drop-down menu

### CODE REVIEW

**Comment by mitar** [comment]
I would like to warn you that you should be sure to commit with same git data as you have on GitHub because otherwise there could be a chance I will miss your contributions (I try not to, but it is harder).

File: piplmesh/frontend/templates/header.html [hunk 1]
```diff
@@ -42,7 +42,49 @@
                         </form>
                     </li>
                 {% else %}
-                    <li><a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a></li>
+                    <li>
+                        <span id="drop_down_login_container">
+                        <a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a>
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Indent properly.

File: piplmesh/frontend/templates/header.html [hunk 2]
```diff
@@ -42,7 +42,49 @@
...
+                        <div id="drop_down_login_options">
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Div inside of span? This is not XHTML valid.

File: piplmesh/frontend/templates/header.html [hunk 3]
```diff
@@ -42,7 +42,49 @@
...
+                                <li>
+                                    <form method="post" action="{% url "google_login" %}" id="google_login_form4">
+                                        {% csrf_token %}
+                                        <p><button type="submit" class="google_login_button"><span class="google_button_text">{% trans "Login with Google" %}</span></button></p>
+                                    </form>
+                                </li>
+                            </ul>
+                        </div></span>
+                        <script type="text/javascript">
+                            $('#drop_down_login_options').hide();
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: No inline JavaScript. Move this to external JS file.

File: piplmesh/frontend/templates/header.html [hunk 4]
```diff
@@ -42,7 +42,49 @@
...
+                            $('#drop_down_login_container').hover(
+                                    function () {
+                                        $('#drop_down_login_options').show();
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: You don't have to have a anonymous wrapper function around, `hower($('#drop_down_login_options').show, $('#drop_down_login_options').hide)` already works. :-)

**Comment by mitar** [comment]
You should improve style ob buttons. Make them of same width and height. No space in between vertically, make them more condensed.

File: piplmesh/frontend/static/piplmesh/js/login_dropdown.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+$('#drop_down_login_options').hide();
+$('#drop_down_login_container').hover(
+    function () {
```
Comment (igypigi) [inline piplmesh/frontend/static/piplmesh/js/login_dropdown.js]: hower($('#drop_down_login_options').show, $('#drop_down_login_options').hide) does not work in my browser :( thats why i changed back to wrapper functions.

**Comment by igypigi** [comment]
I dont understand. Should i commit with name which has #269 in it?

File: piplmesh/frontend/static/piplmesh/js/login_dropdown.js (hunk 5 above)
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/login_dropdown.js]: In which way it does not work? Is there any error displayed?

**Comment by mitar** [comment]
No, you should commit with your real e-mail and not "igor@localhost".

File: piplmesh/frontend/templates/header.html [hunk 6]
```diff
@@ -42,7 +42,40 @@
                         </form>
                     </li>
                 {% else %}
-                    <li><a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a></li>
+                    <li>
+                        <span id="drop_down_login_container">
+                            <a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a>
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Can you make all those buttons not be duplicated? So now we have buttons twice in the templates. Extract this to one template which you then include on both places.

File: piplmesh/frontend/static/piplmesh/js/login_dropdown.js (hunk 5 above)
Comment (igypigi) [inline piplmesh/frontend/static/piplmesh/js/login_dropdown.js]: This is what i get in JS Console:
Napaka: this.animate is not a function
Izvorna datoteka: http://127.0.0.1:8000/static/piplmesh/jquery/jquery.js
Vrstica: 8427

**Comment by igypigi** [comment]
is my email now ok?

File: piplmesh/account/templates/user/other_login_options.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        <p><button type="submit" class="foursquare_login_button"><span class="foursquare_button_text">{% trans "Login with Foursquare" %}</span></button></p>
+    </form>
+</li>
+<li>
+    <form method="post" action="{% url "google_login" %}" id="google_login_form4">
+        {% csrf_token %}
+        <p><button type="submit" class="google_login_button"><span class="google_button_text">{% trans "Login with Google" %}</span></button></p>
+    </form>
+</li>
+<li>{% comment %}Place for OpenID button.{% endcomment %}</li>
```
Comment (mitar) [inline piplmesh/account/templates/user/other_login_options.html]: Can you move also `<li>` into comment?

**Comment by mitar** [comment]
Yes, e-mail is OK now.

File: piplmesh/frontend/templates/header.html [hunk 8]
```diff
@@ -42,7 +42,17 @@
...
+                            <span id="drop_down_login_options">
+                                <ul>
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: You cannot have `<ul>` inside of `<span>`. Please check XHTML compliance.

**Comment by mitar** [comment]
It looks OK now. Please make it XHTML compliant.

**Comment by mitar** [comment]
Please update from main repository.

**Comment by igypigi** [comment]
how do i do this? do i use rebase?

**Comment by igypigi** [comment]
i used some magic and i think it worked... :)

**Comment by mitar** [comment]
No, why would you rebase? You just pull and merge.

**Comment by mitar** [comment]
Hm. Interesting. Which magic have you been using?

File: piplmesh/frontend/templates/header.html [hunk 9]
```diff
@@ -42,7 +42,17 @@
...
+                    <li>
+                        <div id="drop_down_login_container">
+                            <a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a>
+                            <div id="drop_down_login_options">
+                                <ul>
+                                    {% include "user/other_login_options.html" %}
+                                </ul>
+                            </div>
+                        </div>
+                        <script type="text/javascript" src="{% static "piplmesh/js/login_dropdown.js" %}"></script>
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Don't include JavaScript file, add it to the HTML header.

File: piplmesh/frontend/static/piplmesh/js/login_dropdown.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+$('#drop_down_login_options').hide();
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/login_dropdown.js]: You should initially hide with CSS. Not JavaScript.

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 11]
```diff
@@ -581,3 +581,39 @@ input.gsc-search-button {
 .contact_form textarea {
     width: 300px;
 }
+
+/* ------------------ DROP-DOWN LOGIN ---------------------------- */
+#drop_down_login_container {
+    position: relative;
+}
+
+#drop_down_login_container #drop_down_login_options {
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: `#drop_down_login_container #drop_down_login_options` is redundant. Matching by ID is exact. There is no reason why you would match by ID like that. Matching by ID exactly matches that one element. So it cannot happen that there are two elements with same ID, where one would be in a child relation and the other not.

Of course the question is why are you using IDs in the first place. You could use CSS classes.

**Comment by igypigi** [comment]
i could not merge because of conflicts so i just pulled updates to new branch and edited it... and there was some problem pushing it to GitHub so i used some magical flag with git push and it worked :)

**Comment by mitar** [comment]
> i could not merge because of conflicts

But merge conflicts are not bad. They just mean you have to manually resolve them. Manually decide what is the resulting code.

> and there was some problem pushing it to GitHub so i used some magical flag with git push and it worked :)

You lost the history of your commit changes. Which is in practice bad.

**Comment by igypigi** [comment]
is it ok now?

File: piplmesh/frontend/templates/header.html [hunk 12]
```diff
@@ -42,7 +42,16 @@
                         </form>
                     </li>
                 {% else %}
-                    <li><a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a></li>
+                    <li>
+                        <div class="drop_down_login_container">
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Why additional `div`? Why not simply `li` with that class?

File: piplmesh/frontend/templates/header.html [hunk 13]
```diff
@@ -42,7 +42,16 @@
...
+                            <a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a>
+                            <div class="drop_down_login_options">
+                                <ul>
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Same.

**Comment by mitar** [comment]
Please update from main repository. Now we have also support for BrowserID.

File: piplmesh/frontend/templates/base.html [hunk 14]
```diff
@@ -15,6 +15,7 @@
             <script src="{% url "django.views.i18n.javascript_catalog" %}" type="text/javascript"></script>
             <script src="https://browserid.org/include.js" type="text/javascript"></script>
             <script src="{% static "piplmesh/js/common.js" %}" type="text/javascript"></script>
+            <script src="{% static "piplmesh/js/login_dropdown.js" %}" type="text/javascript"></script>
```
Comment (mitar) [inline piplmesh/frontend/templates/base.html:18]: You could move this code to `common.js`.

File: piplmesh/frontend/templates/header.html [hunk 15]
```diff
@@ -42,7 +42,13 @@
                         </form>
                     </li>
                 {% else %}
-                    <li><a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a></li>
+                    <li class="drop_down_login_container">
+                        <a href="{% url "login" %}?{{ REDIRECT_FIELD_NAME }}={{ next_url }}">{% trans "Login" %}</a>
+
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html:47]: Unnecessary empty line.

**Pull request merged**